internal/infra/repository/mongo/favouriteasset/cacherepository.go

```diff
@@ -0,0 +1,95 @@
+package favouriteasset
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	assetsrv "platform-go-challenge/internal/app/favouriteasset"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// CacheTTL contains the time for the result to get cached.
+const CacheTTL = 30 * time.Second
+
+// CRepo is the favouriteasset repo with cache functionality.
+type CRepo struct {
+	repo        Repo
+	redisClient *redis.Client
+}
+
+// NewCacheRepository constructor.
+func NewCacheRepository(repo Repo, redisClient *redis.Client) (*CRepo, error) {
```

Comment:
I think there is a bug in the cache repository (no need to fix it), but how would you write a test to discover it and fix it?

Similarly, if you wanted to upgrade to new mongo DB version, what tests would you add in order to be sure there isn't any issue with the upgrade?

Reply:
Nice catch, first of all, I'm not handling the error case of the Redis FlushDBAsync call. Should create an error channel and fill it into the goroutine. Also, I could make the call run synchronously (without goroutines), but it will have a performance overhead since the program will be blocked during the call.

Secondly, I propagate the repo error directly to the above level, and with this, I will not know if the error is coming from the repo level or the caching level during testing.